publishToConfluence throws exception when uploading attachments in v3.1.0

**Describe the bug**
When uploading attachments, e.g., images, using the new ConfluenceClient in 3.1.0 via the publicToConfluence task, the following error occurs. This affects the usage of v1 and the new v2 Api versions of the ConfluenceClient.
```
> Task :publishToConfluence FAILED

FAILURE: Build failed with an exception.

* Where:
Script '/home/dtcuser/docToolchain/scripts/publishToConfluence.gradle' line: 43

* What went wrong:
Execution failed for task ':publishToConfluence'.
> No such property: MultipartEntityBuilder for class: groovyx.net.http.HTTPBuilder$RequestConfigDelegate

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 26s
4 actionable tasks: 2 executed, 2 up-to-date
```

**To Reproduce**
Steps to reproduce the behavior:
1. Install 3.1.0 docToolChain Docker image
2. Download arc42 template using ./dtcw downloadTemplate
3. Run ./dtcw publishToConfluence ...
4. See error

**Expected behavior**
The publishToConfluence task shouldn't throw an error when uploading attachments.

**Configuration**
 - docToolchain version: 3.1.0
 - OS: MacOS Sonoma
 - Relevant SDK Versions JDK 11
 - Atlassian Confluence: Confluence Cloud

Just an uneducated guess (after skimming the code): in the recent merge https://github.com/docToolchain/docToolchain/pull/1226, the ConfluenceClient seems to miss some import statements, e.g., for MultiPartEntityBuilder and HttpMultipartMode:

https://github.com/docToolchain/docToolchain/blob/cc78e99b0f920a0151e2c5650ebeb2b663908674/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy#L58-L59

--- a/changelog.adoc
+++ b/changelog.adoc
@@ -9,6 +9,8 @@ and this project tries to adhere to https://semver.org/spec/v2.0.0.html[Semantic
 
 === fixes
 
+* https://github.com/docToolchain/docToolchain/issues/1281[#1281: publishToConfluence throws exception when uploading attachments in v3.1.0]
+
 === added
 
 === changed

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy
@@ -1,12 +1,14 @@
 package org.docToolchain.atlassian.clients
 
 import groovyx.net.http.EncoderRegistry
+import groovyx.net.http.HTTPBuilder
 import groovyx.net.http.HttpResponseException
 import groovyx.net.http.Method
 import groovyx.net.http.RESTClient
-import org.apache.http.entity.ContentType
 import org.apache.http.entity.mime.HttpMultipartMode
-import org.apache.http.entity.mime.MultipartEntityBuilder
+import org.apache.http.entity.mime.MultipartEntity
+import org.apache.http.entity.mime.content.InputStreamBody
+import org.apache.http.entity.mime.content.StringBody
 import org.docToolchain.configuration.ConfigService
 
 abstract class ConfluenceClient {
@@ -18,6 +20,8 @@ abstract class ConfluenceClient {
     protected String baseApiUrl
     Map headers
     RESTClient restClient
+    //TODO this is a workaround for the fact that the RESTClient does not support mulitpart/form-data from Groovy 3.0.0 on
+    Map proxyConfig = [:]
 
     ConfluenceClient(ConfigService configService) {
         this.baseApiUrl = configService.getConfigProperty('confluence.api')
@@ -27,6 +31,11 @@ abstract class ConfluenceClient {
         this.editorVersion = determineEditorVersion(configService)
         if(configService.getFlatConfigSubTree('confluence.proxy')){
             def proxy = configService.getFlatConfigSubTree('confluence.proxy')
+            //TODO this is a workaround for the fact that the RESTClient does not support mulitpart/form-data from Groovy 3.0.0 on
+            proxyConfig.put('host', proxy.host as String)
+            proxyConfig.put('port', proxy.port as Integer)
+            proxyConfig.put('schema', proxy.schema  as String?: 'http')
+            //END WORKAROUND
             restClient.setProxy(proxy.host as String, proxy.port as Integer, proxy.schema  as String?: 'http')
         }
         if(configService.getConfigProperty('confluence.bearerToken')){
@@ -56,14 +65,21 @@ abstract class ConfluenceClient {
     abstract createAttachment(String pageId, InputStream inputStream, String fileName, String note, String localHash)
 
     protected uploadAttachment(uri, InputStream inputStream, String fileName, note, localHash) {
-        restClient.request(uri, Method.POST, ContentType.MULTIPART_FORM_DATA){ req ->
-            def multiPartContent = MultipartEntityBuilder.create()
-            .setMode(HttpMultipartMode.BROWSER_COMPATIBLE)
-            .addBinaryBody("file", inputStream, ContentType.APPLICATION_OCTET_STREAM, fileName)
-            .addTextBody("comment", note + "\r\n#" + localHash + "#")
-            .build()
+        def builder = new HTTPBuilder(baseApiUrl + uri)
+        //TODO this is a workaround for the fact that the RESTClient does not support mulitpart/form-data from Groovy 3.0.0 on
+        if (!proxyConfig.isEmpty()) {
+            builder.setProxy(proxyConfig.get("host") as String, proxyConfig.get("port") as Integer, proxyConfig.get("schema") as String)
+        }
+        //END WORKAROUND
+        builder.request(Method.POST) { req ->
+            requestContentType: "multipart/form-data"
+            MultipartEntity multiPartContent = new MultipartEntity(HttpMultipartMode.BROWSER_COMPATIBLE)
+            // Adding Multi-part file parameter "file"
+            multiPartContent.addPart("file", new InputStreamBody(inputStream, fileName))
+            // Adding another string parameter "comment"
+            multiPartContent.addPart("comment", new StringBody(note + "\r\n#" + localHash + "#"))
             req.setEntity(multiPartContent)
-            headers.each { key, value ->
+            this.headers.each { key, value ->
                 req.addHeader(key, value)
             }
         }

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy
@@ -5,6 +5,8 @@ import groovyx.net.http.HttpResponseException
 import groovyx.net.http.Method
 import groovyx.net.http.RESTClient
 import org.apache.http.entity.ContentType
+import org.apache.http.entity.mime.HttpMultipartMode
+import org.apache.http.entity.mime.MultipartEntityBuilder
 import org.docToolchain.configuration.ConfigService
 
 abstract class ConfluenceClient {

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClient.groovy
@@ -64,7 +64,10 @@ abstract class ConfluenceClient {
 
     abstract createAttachment(String pageId, InputStream inputStream, String fileName, String note, String localHash)
 
+    abstract attachmentHasChanged(attachment, localHash)
+
     protected uploadAttachment(uri, InputStream inputStream, String fileName, note, localHash) {
+        //TODO build sane URL
         def builder = new HTTPBuilder(baseApiUrl + uri)
         //TODO this is a workaround for the fact that the RESTClient does not support mulitpart/form-data from Groovy 3.0.0 on
         if (!proxyConfig.isEmpty()) {

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1.groovy
@@ -60,6 +60,12 @@ class ConfluenceClientV1 extends ConfluenceClient {
     }
 
     @Override
+    def attachmentHasChanged(Object attachment, Object localHash) {
+        def remoteHash = attachment.results[0].extensions.comment.replaceAll("(?sm).*#([^#]+)#.*",'$1')
+        return remoteHash!=localHash
+    }
+
+    @Override
     def fetchPagesBySpaceKey(String spaceKey, Integer pageLimit) {
         def allPages = [:]
         Boolean morePages = true

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1.groovy
@@ -55,7 +55,7 @@ class ConfluenceClientV1 extends ConfluenceClient {
 
     @Override
     def createAttachment(String pageId, InputStream inputStream, String fileName, String note, String localHash) {
-        def uri = ""
+        def uri = API_V1_PATH + 'content/' + pageId + '/child/attachment'
         uploadAttachment(uri, inputStream, fileName, note, localHash)
     }
 

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1.groovy
@@ -24,7 +24,7 @@ class ConfluenceClientV1 extends ConfluenceClient {
 
     protected String getRealApiPath() {
         if(this.baseApiUrl.contains("/rest/api")) {
-            def path = new URIBuilder(this.baseApiUrl).getPath()
+            String path = new URIBuilder(this.baseApiUrl).getPath()
             return path.endsWith("/") ? path : path + "/"
         }
         return API_V1_DEFAULT_PATH

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV2.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV2.groovy
@@ -43,7 +43,8 @@ class ConfluenceClientV2 extends ConfluenceClient {
     }
 
     @Override
-    def getAttachment(Object pageId, Object filename) {
+    def getAttachment(Object pageId, Object fileName) {
+        //TODO NOT Found
         restClient.get(
             path: API_V2_DEFAULT_PATH + 'pages/' + pageId + '/attachment',
             query: [
@@ -59,7 +60,7 @@ class ConfluenceClientV2 extends ConfluenceClient {
 
     @Override
     def createAttachment(String pageId, InputStream inputStream, String fileName, String note, String localHash) {
-        def uri = ""
+        def uri = API_V1_DEFAULT_PATH + 'content/' + pageId + '/child/attachment'
         uploadAttachment(uri, inputStream, fileName, note, localHash)
     }
 

--- a/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV2.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV2.groovy
@@ -46,7 +46,7 @@ class ConfluenceClientV2 extends ConfluenceClient {
     def getAttachment(Object pageId, Object fileName) {
         //TODO NOT Found
         restClient.get(
-            path: API_V2_DEFAULT_PATH + 'pages/' + pageId + '/attachment',
+            path: API_V2_DEFAULT_PATH + 'pages/' + pageId + '/attachments',
             query: [
                 'filename': fileName,
             ], headers: headers)
@@ -62,6 +62,12 @@ class ConfluenceClientV2 extends ConfluenceClient {
     def createAttachment(String pageId, InputStream inputStream, String fileName, String note, String localHash) {
         def uri = API_V1_DEFAULT_PATH + 'content/' + pageId + '/child/attachment'
         uploadAttachment(uri, inputStream, fileName, note, localHash)
+    }
+
+    @Override
+    def attachmentHasChanged(attachment, localHash) {
+        def remoteHash = attachment.results[0].comment.replaceAll("(?sm).*#([^#]+)#.*",'$1')
+        return remoteHash!=localHash
     }
 
     @Override

--- a/core/src/main/groovy/org/docToolchain/configuration/ConfigService.groovy
+++ b/core/src/main/groovy/org/docToolchain/configuration/ConfigService.groovy
@@ -16,11 +16,11 @@ class ConfigService {
     }
 
     Map getFlatConfigSubTree(String propertyPath) {
-       return config.flatten().inject([:]) { result, key, value-> {
+       return config.flatten().inject([:]) { result, key, value ->
             if(key.startsWith(propertyPath)){
                 result.put(key.replaceFirst("${propertyPath}.", ""), value)
             }
             return result
-        }}
+        }
     }
 }

--- a/core/src/main/groovy/org/docToolchain/scripts/asciidoc2confluence.groovy
+++ b/core/src/main/groovy/org/docToolchain/scripts/asciidoc2confluence.groovy
@@ -138,10 +138,9 @@ def uploadAttachment = { def pageId, String url, String fileName, String note ->
     }
 
     def attachment = confluenceClient.getAttachment(pageId, fileName).data
-    if (attachment.size==1) {
+    if (attachment.size()>0 && attachment.results.size()>0) {
         // attachment exists. need an update?
-        def remoteHash = attachment.results[0].extensions.comment.replaceAll("(?sm).*#([^#]+)#.*",'$1')
-        if (remoteHash!=localHash) {
+        if (confluenceClient.attachmentHasChanged(attachment, localHash)) {
             //hash is different -> attachment needs to be updated
             confluenceClient.updateAttachment(pageId, attachment.results[0].id, is, fileName, note, localHash)
             println "    updated attachment"

--- a/core/src/test/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1Spec.groovy
+++ b/core/src/test/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV1Spec.groovy
@@ -75,4 +75,27 @@ class ConfluenceClientV1Spec extends ConfluenceClientSpec {
         then: "the default API path is set correctly"
             apiPath == "/wiki/rest/api/"
     }
+
+    def "test attachment diff can be detected correctly"() {
+        def mockedResponse =  [results:[
+            [id:"att178445677", type:"attachment", status:"current", title:"01_2_iso-25010-topics-DE.drawio.png",
+             macroRenderedOutput:[:], metadata:[comment:"automatically uploaded #a937de4e4d3a556d0b8886d7ecb5e0c5#", mediaType:"image/png"],
+             extensions:[mediaType:"image/png", fileSize:"100109", comment:"automatically uploaded #a937de4e4d3a556d0b8886d7ecb5e0c5#",
+                         mediaTypeDescription:"PNG Image", fileId:"e790d35e-bdc1", collectionName:"contentId-44444"],
+             _expandable:[], _links:[webui:"", self:"", download:""]]], start:0, limit:50, size:1, _links:[base:"", context:"/wiki", self:""]]
+        when: "i create a ConfluenceClientV1"
+            ConfigObject config = new ConfigObject()
+            config.confluence = [
+                api: "https://confluence.atlassian.com",
+                credentials: "user:password"
+            ]
+            ConfigService configService = new ConfigService(config)
+            ConfluenceClient confluenceClient = getConfluenceClient(configService)
+        and : "i check if the attachment has changed"
+            def hasNotChanged = confluenceClient.attachmentHasChanged(mockedResponse, "a937de4e4d3a556d0b8886d7ecb5e0c5")
+            def hasChanged = confluenceClient.attachmentHasChanged(mockedResponse, "somethingdifferent")
+        then: "the attachment has changed"
+            hasChanged == true
+            hasNotChanged == false
+    }
 }

--- a/core/src/test/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV2Spec.groovy
+++ b/core/src/test/groovy/org/docToolchain/atlassian/clients/ConfluenceClientV2Spec.groovy
@@ -58,4 +58,33 @@ class ConfluenceClientV2Spec extends ConfluenceClientSpec {
             confluenceClient != null
             confluenceClient.spaceId == "123456"
     }
+
+    def "test attachment diff can be detected correctly"() {
+        def mockedResponse =   [results:[[mediaTypeDescription:"PNG Image",
+                                          webuiLink:"", downloadLink:"", createdAt:"null", id:"att17858747",
+                                          comment:"automatically uploaded #a937de4e4d3a556d0b8886d7ecb5e0c5#",
+                                          version: [number:1,
+                                                    message:"automatically uploaded #a937de4e4d3a556d0b8886d7ecb5e0c5#",
+                                                    minorEdit:false, authorId:"557058:5d896875-errew-85a4-werwerwerwerwe",
+                                                    createdAt:2
+                                          ],
+                                          title:"01_2_iso-25010-topics-DE.drawio.png", fileSize:100109, status:"current",
+                                          mediaType:"image/png", pageId:17695008, fileId:"e79-rtzrtz-tzrz-d7aebea2986f",
+                                          _links:[
+                                              download:"", webui:""]]], _links:[:]]
+        when: "i create a ConfluenceClientV2"
+            ConfigObject config = new ConfigObject()
+            config.confluence = [
+                api: "https://confluence.atlassian.com",
+                credentials: "user:password"
+            ]
+            ConfigService configService = new ConfigService(config)
+            ConfluenceClient confluenceClient = getConfluenceClient(configService)
+        and : "i check if the attachment has changed"
+            def hasNotChanged = confluenceClient.attachmentHasChanged(mockedResponse, "a937de4e4d3a556d0b8886d7ecb5e0c5")
+            def hasChanged = confluenceClient.attachmentHasChanged(mockedResponse, "somethingdifferent")
+        then: "the attachment has changed"
+            hasChanged == true
+            hasNotChanged == false
+    }
 }

--- a/scripts/publishToConfluence.gradle
+++ b/scripts/publishToConfluence.gradle
@@ -22,7 +22,9 @@ task verifyConfluenceApiAccess(
 ) {
     doLast {
         //TODO default should be false, if the V1 has been removed in cloud
-        config.confluence.useV1Api = findProperty("confluence.useV1Api")?:config.confluence.useV1Api
+        config.confluence.useV1Api = findProperty("confluence.useV1Api") != null ?
+            findProperty("confluence.useV1Api") : config.confluence.useV1Api != [:] ?
+            config.confluence.useV1Api :true
         new VerifyConfluenceApiAccessTask(config).execute()
     }
 }
@@ -37,7 +39,9 @@ task publishToConfluence(
         logger.info("docToolchain> docDir: "+docDir)
         config.confluence.api = findProperty("confluence.api")?:config.confluence.api
         //TODO default should be false, if the V1 has been removed in cloud
-        config.confluence.useV1Api = (findProperty("confluence.useV1Api")?:config.confluence.useV1Api)
+        config.confluence.useV1Api = findProperty("confluence.useV1Api") != null ?
+            findProperty("confluence.useV1Api") : config.confluence.useV1Api != [:] ?
+            config.confluence.useV1Api :true
         binding.setProperty('config',config)
         binding.setProperty('docDir',docDir)
         evaluate(new File(projectDir, 'core/src/main/groovy/org/docToolchain/scripts/asciidoc2confluence.groovy'))

--- a/scripts/publishToConfluence.gradle
+++ b/scripts/publishToConfluence.gradle
@@ -22,7 +22,7 @@ task verifyConfluenceApiAccess(
 ) {
     doLast {
         //TODO default should be false, if the V1 has been removed in cloud
-        config.confluence.useV1Api = findProperty("confluence.useV1Api")?:config.confluence.useV1Api?:true
+        config.confluence.useV1Api = findProperty("confluence.useV1Api")?:config.confluence.useV1Api
         new VerifyConfluenceApiAccessTask(config).execute()
     }
 }
@@ -37,7 +37,7 @@ task publishToConfluence(
         logger.info("docToolchain> docDir: "+docDir)
         config.confluence.api = findProperty("confluence.api")?:config.confluence.api
         //TODO default should be false, if the V1 has been removed in cloud
-        config.confluence.useV1Api = (findProperty("confluence.useV1Api")?:config.confluence.useV1Api)?:true
+        config.confluence.useV1Api = (findProperty("confluence.useV1Api")?:config.confluence.useV1Api)
         binding.setProperty('config',config)
         binding.setProperty('docDir',docDir)
         evaluate(new File(projectDir, 'core/src/main/groovy/org/docToolchain/scripts/asciidoc2confluence.groovy'))

--- a/src/docs/015_tasks/03_task_publishToConfluence.adoc
+++ b/src/docs/015_tasks/03_task_publishToConfluence.adoc
@@ -91,6 +91,9 @@ If you use Confluence Server, you may need to set a context, depending on your C
 
 *_This feature is available for docToolchain >= 3.1 only_*
 
+[NOTE]
+If you set this to `false`, ensure the `api` config is set to `https://[yourCloudDomain]`. (Mind no _context_ given here)
+
 If you are using Confluence Cloud, you can set this to `false` to use the new API V2. If you are using Confluence Server, you can set this to `true` to use the old API V1. If you are using Confluence Cloud and set this to `false`, you will get an error message, once Atlassian turns off API V1 (starting 01.01.2024).
 
 *enforceNewEditor*

